### ISSUE TITLE
Affinitizing GC threads to available cpuset in containers

### ISSUE THREAD

**Issue by ezsilmar** [dotnet/coreclr#25875]
Hello,

After optimizing GC for a [small container scenario](https://devblogs.microsoft.com/dotnet/running-with-server-gc-in-a-small-container-scenario-part-1-hard-limit-for-the-gc-heap/) (commit ed52a006c01a582d4d34add40c318d6f324b99ba) GC threads are not affinitized by default if the app runs with a memory limit. This is fine in most cases but is not convenient if you use cpuset (instead of more default cfs_bandwidth) isolation. 

By cpuset isolation I mean a case when the container is allowed to use only specific cpu cores. For example, if an app asks for 5 cpus, the isolator will pin the app to some free cores (f.e. 0-2,12,18) and guarantee that nobody else is scheduled on these cpus.

In this case, we can certainly benefit from the gc threads affinitizing. With 2.x runtime it works out of the box and gc threads are pinned correctly. However, with 3.0 we need to explicitly pass `COMPlus_GCHeapAffinitizeMask` to activate the feature. This requires the isolator to pass an affinity mask to the application which will unnecessary couple them tighter.

It would be great if we could just set some flag to say that GC should affinitize its threads to allowed cpuset, the same way we can set `COMPlus_GCNoAffinitize` to 1 in 2.x to say it shouldn't.

I also discovered that the value of `COMPlus_GCHeapAffinitizeMask` is actually ignored in this case. If it's set to any value other than 0 GC threads will be affinitized as in 2.x (resulting in the behavior I want). Not sure though if this behavior was intended or it's a bug?

**Comment by Maoni0** [comment]
> I also discovered that the value of COMPlus_GCHeapAffinitizeMask is actually ignored in this case. If it's set to any value other than 0 GC threads will be affinitized as in 2.x (resulting in the behavior I want). Not sure though if this behavior was intended or it's a bug?

if you did specify affinitymask we will use that as your intention to affinitze so this is by design. 

as long as we can know which cores the container is affinitized to, we can have GC automatically affinitize threads on them. @janvorli would you have time to work on this?

**Comment by ezsilmar** [comment]
> as long as we can know which cores the container is affinitized to, we can have GC automatically affinitize threads on them.

It already works that way if you pass `COMPlus_GCHeapAffinitizeMask=1`. I meant it's a bit confusing that `GCHeapAffinitizeMask` value is ignored. F.e. imagine cpuset restricts the app to `0,1,2,3,4` and you set the mask to `12` (meaning `2,3`). You expect the app to have 2 gc heaps and 2 gc threads affinitized to `2,3`. Instead there will be 5 heaps with threads pinned to `0-4`.

However, I can't come up with a real-world usage of that scenario (when, in a container, you want to pin gc threads to something different from allowed cpuset). Current behavior looks good to me if it's documented.

**Comment by Maoni0** [comment]
> imagine cpuset restricts the app to 0,1,2,3,4 and you set the mask to 12 (meaning 2,3). You expect the app to have 2 gc heaps and 2 gc threads affinitized to 2,3. Instead there will be 5 heaps with threads pinned to 0-4.

that should not be the behavior; if this is what happens it's a bug. @janvorli could you please see if you could repro this behavior?

`uint32_t num_affinitized_processors = (uint32_t)process_affinity_set->Count();`

that should return 2 if you specify 12 as mask (just making sure if you meant you specified 12 as the value of this env var, it means 0x12 so it should still be affinitizing to 2 cores but different cores).

**Comment by janvorli** [comment]
The `COMPlus_GCHeapAffinitizeMask` is ignored on Unix in 3.0 for setting affinity and only `COMPlus_GCHeapAffinitizeRanges` can be used to set a requested affinity. However, I can see that even though the `COMPlus_GCHeapAffinitizeMask` is ignored on Unix, we still set `affinity_config_specified_p` to `true` if it is set. That causes the GC heaps to get affinitized to all available CPUs. This is something we should fix.
@ezsilmar you should use the `COMPlus_GCHeapAffinitizeRanges` on Unix instead. The format is a comma separated list of ranges of CPU indices (e.g. `0-3,5,6-7`). So if you want to affinitize to CPUs 2 and 3 as you've mentioned, you can set it either to `2-3` or `2,3`, whatever you prefer.

**Comment by ezsilmar** [comment]
Wow, I didn't know about `COMPlus_GCHeapAffinitizeRanges` parameter. So the `COMPlus_GCHeapAffinitizeMask` is deprecated since 3.0.0?

> even though the `COMPlus_GCHeapAffinitizeMask` is ignored on Unix, we still set `affinity_config_specified_p` to `true` if it is set. That causes the GC heaps to get affinitized to all available CPUs. This is something we should fix.

After you fix that, what will be the correct way to configure "affinitize GC threads to all available CPUs" in containers? Explicitly passing the affinity mask is not convenient, as I explained in the issue description.

One way I see is to check if there're both a `cpuset` restriction and a memory limit. If both are present, affinitize to all CPUs by default, as in no-container case. It won't affect the usual case with `cfs_bandwidth` isolation and one can still override this behavior by setting `COMPlus_GCNoAffinitize` to `1`. However I'm not sure we can easily check if `cpuset` restriction applies to our process from the inside.

**Comment by janvorli** [comment]
> So the `COMPlus_GCHeapAffinitizeMask` is deprecated since 3.0.0?

No, but it is supported only on Windows and only when CPU groups are not enabled (that means multiple CPU groups will be used). The mask is relative to the CPU group the current process is running in. Since Unix has no concept of CPU groups (it is a windows way of handling more than 64 CPUs while still making some old APIs that predated support for more than 64 CPUs in Windows work).


> what will be the correct way to configure "affinitize GC threads to all available CPUs" in containers?

Even without a fix to that minor overlook, you can configure that by setting the `COMPlus_GCHeapAffinitizeMask` to `0-1023`. The specified mask can be larger than the real number of processors on the device and 0-1023 effectively means "all". You can also set the upper limit to the number of processors on your Linux device instead of the hardcoded 1023.
The mask is applied the following way:
* We read the current process affinity, which basically gets the set of cpus applied by the `cpuset`.
* Then we remove all processors that were not specified by the heap affinitize mask.
* Now we have a set of processors that GC uses for creating GC heaps / affinitizing GC threads.

**Comment by Maoni0** [comment]
oh right, I forgot that on Linux the mask config is ignored. @janvorli will you be making the fix?

> you can configure that by setting the COMPlus_GCHeapAffinitizeMask to 0-1023. 

typo...  should be COMPlus_GCHeapAffinitizeRanges.

**Comment by janvorli** [comment]
Thanks @Maoni0 for correcting the typo. Correcting my comment.

**Comment by janvorli** [comment]
> @janvorli will you be making the fix?

Sure, I can make it.

**Comment by Maoni0** [comment]
great, thanks @janvorli!
